### PR TITLE
feat(eks): control plane logs

### LINKED ISSUE #4159: [aws-eks] Enable Control Plane logs in EKS cluster
### Use Case

Enabling Control Plane logging in EKS cluster is only possible by calling EKS API after cluster is created. Doing it in CDK requires to create Custom Resource with code that calls the API. It would be nice to have it as an argument for creating EKS cluster from CDK.

### Proposed Solution

Since EKS is created from python lambda when kubectlEnabled flag is enabled there is a simple way to create the EKS cluster with logging enabled. Currently the lambda code uses boto3 method eks.create_cluster() where we can pass arguments to enable logging on created cluster. (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/eks.html#EKS.Client.create_cluster).
The lambda uses config as an argument for this method :

https://github.com/aws/aws-cdk/blob/c3b3c935b2086288dd79f559a49c3b5304a8e40f/packages/%40aws-cdk/aws-eks/lib/cluster-resource/index.py#L69

The config is passed as a properties of custom resource and is created here: 

https://github.com/aws/aws-cdk/blob/c3b3c935b2086288dd79f559a49c3b5304a8e40f/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L364-L379

So I suggest to expose a way to include logging properties in the config so it should be passed to eks.create_cluster() method without any more changes. That should result in enabling logging on newly created EKS cluster.

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by eduardomourar**
This creates an option to enable the control plane logs during creation of EKS cluster. For more information, see [documentation here](https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html).

Due to a limitation in the CloudFormation custom resources, we will allow booleans values to be encoded into special strings (`true` into `'TRUE:BOOLEAN'` and `false` into `'FALSE:BOOLEAN'`).

closes #4159

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/custom-resource.ts [hunk 1]
```diff
@@ -91,6 +91,13 @@ export interface CustomResourceProps {
    * @default false
    */
   readonly pascalCaseProperties?: boolean;
+
+  /**
+   * Encode property values from boolean to special strings.
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/custom-resource.ts:96]:
```suggestion
   * Encode property values from to special strings:
   * - Booleans are encoded to `TRUE:BOOLEAN` and `FALSE:BOOLEAN`
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts [hunk 2]
```diff
@@ -14,6 +14,22 @@ export interface EksUpdateId {
 
 export type ResourceEvent = AWSLambda.CloudFormationCustomResourceEvent & EksUpdateId;
 
+/**
+ * Decodes encoded true/false values
+ */
+function decodeBooleans(object: object) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts:20]: Why is this needed if we have this in the custom resource library?

**Comment by eladb** [comment]
Ping?

**Comment by eduardomourar** [comment]
pong. sorry i will continue working on this

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts (hunk 2 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts]: you are right. removed

**Comment by eduardomourar** [comment]
> If we want to implement this at a low level, I think that `core.CustomResource` should inject a special property to the custom resource which indicates that values are encoded and then automatically decode these automatically in `core.CustomResourceProvider` and `custom-resources.Provider`.
> 
> If this is too much, I am okay with just doing this at the EKS level for now.

I believe I covered all the runtime codes that you mentioned. Please, have a look.

**Comment by eduardomourar** [comment]
I am unable to figured it out how to solve the JSII error:
> error TS0: Exported APIs cannot use un-exported type @aws-cdk/aws-eks.ControlPlaneLogging

**Comment by eduardomourar** [comment]
BTW, is there a way to load the interfaces (like the control plane logging) from `aws-sdk`? Otherwise we have to duplicate them in `@aws-cdk/aws-eks`, which seems pointless.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 3]
```diff
@@ -168,6 +168,13 @@ export interface ClusterOptions {
    */
   readonly version?: string;
 
+  /**
+   * Enable or disable exporting the Kubernetes control plane logs for your cluster to CloudWatch Logs.
+   *
+   * @default - False, meaning that cluster control plane logs are not exported to CloudWatch Logs.
+   */
+  readonly controlPlaneLogging?: boolean | ControlPlaneLogging;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:176]: Union types cannot be translated to certain languages and therefore not allowed in the CDK APIs.

Here's the pattern we have for this use case:

```ts
readonly controlPlaneLogging?: boolean;
readonly controlPlaneLoggingOptions?: ControlPlaneLogging;
```

- if `controlPlaneLogging` is `true` and `options` is undefined, then it will enable all logs.
- if `controlPlaneLoggingOptions` has a value (some config) and `controlPlaneLogging` is undefined, the it enables control plane logging with the specific config.
- if `controlPlaneLoggingOptions` has a value and `controlPlaneLogging` is explicitly `false` - raise an error.

This patterns offers an ergonomic experience:

```ts
// enable all logs (common use case)
new Cluster(this, 'cluster', { controlPlaneLogging: true });

// enable certain logs
new Cluster(this, 'cluster', {
  controlPlaneLoggingOptions: { ... }
});
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource.ts [hunk 4]
```diff
@@ -4,6 +4,49 @@ import { CLUSTER_RESOURCE_TYPE } from './cluster-resource-handler/consts';
...
+/**
+ * EKS cluster control plane logging configuration
+ */
+export interface ControlPlaneLogging {
+
+  /**
+   * The cluster control plane logging configuration for your cluster.
+   */
+  clusterLogging: Array<ClusterLogging>;
+}
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource.ts:34]: This API seems unnecessarily convoluted. Can we use a boolean map?

```ts
export interface ControlPlaneLogging {
  readonly api?: boolean;
  readonly audit?: boolean;
  readonly authenticator?: boolean;
  readonly controllerManager?: boolean;
  readonly scheduler?: boolean;
}
```

File: packages/@aws-cdk/aws-eks/lib/cluster-resource.ts [hunk 5]
```diff
@@ -4,6 +4,49 @@ import { CLUSTER_RESOURCE_TYPE } from './cluster-resource-handler/consts';
...
+  /**
+   * The available cluster control plane log types.
+   */
+  readonly types: Array<'api' | 'audit' | 'authenticator' | 'controllerManager' | 'scheduler'>;
+}
+
+/**
+ * EKS cluster control plane logging configuration
+ */
+export interface ControlPlaneLogging {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource.ts:28]: This file is internal, so if you want this type to be used in the public API, you'll need to move it to a different file that is exported in `index.ts`

File: packages/@aws-cdk/aws-eks/README.md [hunk 6]
```diff
@@ -386,6 +386,38 @@ A convenience method for mapping a role to the `system:masters` group is also av
...
+By default, cluster control plane logs aren't sent to CloudWatch Logs. You must
+enable each log type individually to send logs for your cluster. You can activate
+all control plane logs by having the property `controlPlaneLogging: true`,
+then the following setting will be applied:
+
+```ts
+new eks.Cluster(this, 'Cluster', {
+  kubectlEnabled: true,
+  controlPlaneLogging: {
+    clusterLogging: [
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/README.md:404]: see notes about API

File: packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts [hunk 7]
```diff
@@ -21,9 +21,27 @@ export type HandlerResponse = undefined | {
   NoEcho?: boolean;
 };
 
+/**
+ * Decodes encoded true/false values
+ */
+function decodeBooleans(object: object) {
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts:27]: decode should only be called if `encodeValues` is enabled. I believe we will need to pass this information to the custom resource provider some how.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 3 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:176]: ok, so i will make the changes as suggested

File: packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts (hunk 7 above)
Comment (eduardomourar) [reply packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts:27]: i am supposed to move the `decodeBooleans` only to EKS, right? or do you want to do it only for the `custom-resources.Provider` and `custom-resources.AwsCustomResource`?

Comment (eladb) [reply packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts:27]: Let's just keep this change only in the EKS module if that's possible.

Comment (eduardomourar) [reply packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts:27]: after starting to remove everything from the other libraries, i realized that by using the environment variables from the lambda themselves we can minimize the impact of the changes. see if we can still follow with the initial low-level approach like this, otherwise i will revert everything back

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 3 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: done. btw, thanks for the explanation!

File: packages/@aws-cdk/aws-eks/README.md [hunk 8]
```diff
@@ -386,6 +386,37 @@ A convenience method for mapping a role to the `system:masters` group is also av
...
+to CloudWatch Logs in your account.
+
+By default, cluster control plane logs aren't sent to CloudWatch Logs. You must
+enable each log type individually to send logs for your cluster. You can activate
+all control plane logs by having the property `controlPlaneLogging: true`,
+then the following setting will be applied:
+
+```ts
+new eks.Cluster(this, 'Cluster', {
+  kubectlEnabled: true,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/README.md:402]: This is the default:

```suggestion
```

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 9]
```diff
@@ -168,6 +169,14 @@ export interface ClusterOptions {
    */
   readonly version?: string;
 
+  /**
+   * Enable or disable exporting the Kubernetes control plane logs for your cluster to CloudWatch Logs.
+   *
+   * @default - False, meaning that cluster control plane logs are not exported to CloudWatch Logs.
+   */
+  readonly controlPlaneLogging?: boolean;
+  readonly controlPlaneLoggingOptions?: ControlPlaneLogging;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:178]: Missing docstring

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py [hunk 10]
```diff
@@ -24,7 +41,7 @@ def helm_handler(event, context):
     release      = props['Release']
     chart        = props['Chart']
     version      = props.get('Version', None)
-    wait         = props.get('Wait', False)
+    wait         = decode_boolean(props.get('Wait', False))
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py:44]: Why is this needed? was this a bug??

File: packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  readonly api?: boolean;
+  readonly audit?: boolean;
+  readonly authenticator?: boolean;
+  readonly controllerManager?: boolean;
+  readonly scheduler?: boolean;
+}
+
+/**
+ * The available cluster control plane log types.
+ */
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts:14]: This type should not be exported from the module. I think it can just go into cluster-resource.ts

File: packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+/**
+ * EKS cluster control plane logging configuration
+ */
+export interface ControlPlaneLogging {
+  readonly api?: boolean;
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts:5]: Missing docstrings

File: packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+/**
+ * EKS cluster control plane logging configuration
+ */
+export interface ControlPlaneLogging {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts:4]: Can we move this type to cluster.ts?

File: packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts (hunk 7 above)
Comment (eladb) [reply packages/@aws-cdk/core/lib/custom-resource-provider/nodejs-entrypoint.ts:28]: Let’s revert please.

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py (hunk 10 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py:44]: yes, it was

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py:44]: I guess the bug was when it was explicitly set to “false” and evaluated as truthy. If it was not set it evaluates as False, which is fine and the default behavior, and If “Wait” was “true” it still evaluates as truthy. 

If that’s the case, I rather just ensure that Wait is assigned an empty string or undefined from the construct when it is false instead of playing with decoding.

File: packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts (hunk 13 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts:4]: i guess not, because we are using in both cluster.ts and cluster-resource.ts

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/shared-interfaces.ts:4]: Okay so let’s rename this file to `logging.ts` please

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py (hunk 10 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/helm/__init__.py:44]: alright

File: packages/@aws-cdk/aws-cloudformation/test/integ.core-custom-resources.expected.json [hunk 14]
```diff
@@ -27,7 +27,7 @@
       "Properties": {
         "Code": {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudformation/test/integ.core-custom-resources.expected.json:28]: Revert

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-provider.ts [hunk 15]
```diff
@@ -56,6 +56,7 @@ export class ClusterResourceProvider extends NestedStack {
       isCompleteHandler: isComplete,
       totalTimeout: Duration.hours(1),
       queryInterval: Duration.minutes(1),
+      isValueEncoded: true,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-provider.ts:59]: Revert

File: packages/@aws-cdk/aws-eks/lib/helm-chart.ts [hunk 16]
```diff
@@ -93,13 +93,14 @@ export class HelmChart extends Construct {
     new CustomResource(this, 'Resource', {
       serviceToken: provider.serviceToken,
       resourceType: HelmChart.RESOURCE_TYPE,
+      encodeValues: true,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/helm-chart.ts:96]: Revert

File: packages/@aws-cdk/aws-eks/lib/helm-chart.ts [hunk 17]
```diff
@@ -93,13 +93,14 @@ export class HelmChart extends Construct {
...
       properties: {
         ClusterName: props.cluster.clusterName,
         RoleArn: props.cluster._kubectlCreationRole.roleArn,
         Release: props.release ?? this.node.uniqueId.slice(-53).toLowerCase(), // Helm has a 53 character limit for the name
         Chart: props.chart,
         Version: props.version,
-        Wait: props.wait ?? false,
+        Wait: props.wait || undefined,
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/helm-chart.ts:103]: Add a comment explaining why this needs to be “undefined” to represent false

File: packages/@aws-cdk/aws-eks/lib/utils.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Encodes values (specially boolean) as special strings
+ *
+ * Because CloudFormation converts every input as string for custom resources,
+ * we will use these encoded values to cast them into proper types.
+ */
+export function encodeValues(object: object) {
+  return JSON.parse(JSON.stringify(object), (_k, v) => {
+    switch (v) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/utils.ts:9]: Revert

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts [hunk 19]
```diff
@@ -14,6 +14,22 @@ export interface EksUpdateId {
 
 export type ResourceEvent = AWSLambda.CloudFormationCustomResourceEvent & EksUpdateId;
 
+/**
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts:17]: Revert

File: packages/@aws-cdk/aws-iam/test/integ.condition-with-ref.expected.json [hunk 20]
```diff
@@ -4,17 +4,17 @@
...
-    "AssetParameters3b28f4ee261986c158a160900e3042a61238f644fe502199d60bcea592128086S3VersionKey4BC65AD6": {
+    "AssetParameters7e4a4a22006fdd286201df6f10f818aa811147ef55011769ef7508309697ea4bS3VersionKey3A27CEA6": {
       "Type": "String",
-      "Description": "S3 key for asset version \"3b28f4ee261986c158a160900e3042a61238f644fe502199d60bcea592128086\""
+      "Description": "S3 key for asset version \"7e4a4a22006fdd286201df6f10f818aa811147ef55011769ef7508309697ea4b\""
     },
-    "AssetParameters3b28f4ee261986c158a160900e3042a61238f644fe502199d60bcea592128086ArtifactHashD8D99435": {
+    "AssetParameters7e4a4a22006fdd286201df6f10f818aa811147ef55011769ef7508309697ea4bArtifactHash2350D6E2": {
       "Type": "String",
-      "Description": "Artifact hash for asset \"3b28f4ee261986c158a160900e3042a61238f644fe502199d60bcea592128086\""
```
Comment (eladb) [inline packages/@aws-cdk/aws-iam/test/integ.condition-with-ref.expected.json:17]: Revert

File: packages/@aws-cdk/core/lib/custom-resource-provider/consts.ts [hunk 21]
```diff
@@ -0,0 +1 @@
+export const IS_VALUE_ENCODED_ENV = 'USER_IS_VALUE_ENCODED';
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/custom-resource-provider/consts.ts:1]: Revert

File: packages/@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider.ts [hunk 22]
```diff
@@ -56,6 +57,13 @@ export interface CustomResourceProviderProps {
    * @default Size.mebibytes(128)
    */
   readonly memorySize?: Size;
+
+  /**
+   * Whether we should decode property values (such as booleans) from special strings.
```
Comment (eladb) [inline packages/@aws-cdk/core/lib/custom-resource-provider/custom-resource-provider.ts:62]: Revert

**Comment by eduardomourar** [comment]
reverting everything back is taking a little bit of time, but it will be done

**Comment by eduardomourar** [comment]
> What happens if logging Config is updated? Have you tested this scenario?

I have tested it (probably 2 weeks ago), but I will do it again after the automated testing is green

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts (hunk 19 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/common.ts:17]: this is actually needed because the provide framework is not be handling it anymore

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 23]
```diff
@@ -244,6 +245,21 @@ export interface ClusterProps extends ClusterOptions {
    */
   readonly kubectlEnabled?: boolean;
 
+  /**
+   * Enable or disable exporting the Kubernetes control plane logs for your cluster to CloudWatch Logs.
+   *
+   * @default false Cluster control plane logs are not exported to CloudWatch Logs.
+   */
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: What happens if I enabled logs and then set to undefined (remove it). Would it set all the logging to false?

Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: it will send an empty value for logging in the call to `updateClusterConfig()`, and it would result in setting the default value from AWS

File: packages/@aws-cdk/aws-eks/README.md [hunk 24]
```diff
@@ -386,6 +386,36 @@ A convenience method for mapping a role to the `system:masters` group is also av
...
+    scheduler: true,
+  }
+});
+```
+
+For more details about the log types available check the
+[API reference documentation](https://docs.aws.amazon.com/eks/latest/APIReference/API_LogSetup.html).
+
+**NOTE**: the control plane logging is only available if the cluster
+has the `kubectlEnabled: true`, because there is still no
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/README.md:416]:
```suggestion
has the `kubectlEnabled: true` (which is the default), because there is still no
```

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 23 above)
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: Have you tested this? My intuition is that if I send an `updateClusterConfig` request with an empty value for `logging` it will not perform any changes to my logging configuration, which is *not* what users expect if they enable logs and then remove this property in their CDK app.

I suspect that we may either need to *always* send the two enabled/disable lists *or* add heuristics to the UPDATE handler to calculate the "diff" from the previous value and produce an appropriate update.

Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: i will try this out and let you know, but this diff is being done in the runtime code here: https://github.com/aws/aws-cdk/blob/53d647f25d5ad92817eb8d0773b5d1af9b95a96a/packages/%40aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts#L293

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: This only determines that there was a change, not which log types should now be enabled/disabled based on that change.

For example, say, "oldProps" included two enabled types (say `["typeA", "typeB"]`) and now "newProps" include only a single type (say `["typeB"]`). Then, the request we need to send to `updateClusterConfig` is only to __disable__ (we can, if we want also include an "enable" for `typeB`, but technically it's not required because it is already enabled).

LMK if this makes sense.

Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:252]: i believe we will need to send the whole configuration explicitly, including the disabled one. i am getting the following error when sending the exact same configuration to `updateClusterConfig`:
> Failed to update resource. Error: No changes needed for the logging config provided

**Comment by eduardomourar** [comment]
I do think this is a lot more complex than I expected, so I would rather abandon this PR (or the CDK can just pick from where I left off). Just a few things to consider:

- The `updateClusterConfig` can have its own custom resource so that we have the control plane logging and the access endpoint (#5220) configuration decoupled from cluster creation process.
- The boolean encoding/decoding is a recurring issue, and I see people are avoiding non strings in the custom resource providers being implemented because of that. So it should be somehow addressed from the CDK core and I am happy to contribute to it if needed be.

File: packages/@aws-cdk/aws-eks/lib/helm-chart.ts (hunk 17 above)
Comment (eduardomourar) [reply packages/@aws-cdk/aws-eks/lib/helm-chart.ts:105]: @eladb , do you want me to create a separate PR just to address this bug with the `wait` parameters in the helm charts?

**Comment by eladb** [comment]
@eduardomourar the helm boolean flags for helm charts are being fixed by @pahud as part of #8787

Do we have other flags?

**Comment by eduardomourar** [comment]
I don’t believe we currently have any other flag, unless we introduce more parameters for the helm command.

**Comment by adriantaut** [comment]
@eladb I have a slightly different approach that I've just implemented in our company. Would like to get your pre-approval on the approach before preparing a PR, if this is possible. 

```ts
import { AwsCustomResource, AwsCustomResourcePolicy } from "@aws-cdk/custom-resources";

const SUPPORTED_CONTROL_PLANE_LOGGING = ['api', 'audit', 'authenticator', 'controllerManager', 'scheduler'] as const;
type SUPPORTED_CONTROL_PLANE_LOGGING = typeof SUPPORTED_CONTROL_PLANE_LOGGING[number];

export interface EksControlPlaneLoggingProps {
  /**
   * Control Plane Logging according to https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html
   * If the `types` parameter is not explicitly provided, all the available Control Plane logging types will be enabled.
   *
   * @default false
   */
  readonly enabled: boolean;
  /**
   * Allows the enablement of only specific Control Plane Logging types
   */
  readonly types?: SUPPORTED_CONTROL_PLANE_LOGGING[];
}


export interface EksClusterProps {
.............................
  /**
   * EKS Control Plane Logging Configuration
   */
  readonly controlPlaneLogging?: EksControlPlaneLoggingProps,
.............................
}


/**
 * Base PPB Class for creating EKS CLusters
 */
export class EksCluster extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: EksClusterProps) {
    super(scope, id);

    .............................

    // instantiate the EKS Cluster
    this.cluster = this.createCluster(props, this.vpc);

    // add the NodeGroup according to user specifications
    this.nodeGroup = this.addNodeGroup(props.nodeGroup);

    // enable EKS Control Plane Logging if defined by user
    this.enableControlPlaneLogging(props.controlPlaneLogging);
    .............................
  }

  private enableControlPlaneLogging(props?: EksControlPlaneLoggingProps) {    
    if (props?.enabled) {
      const enabled_types = props.types ?? SUPPORTED_CONTROL_PLANE_LOGGING;
      const disabled_types = SUPPORTED_CONTROL_PLANE_LOGGING.filter(item => enabled_types.indexOf(item) < 0);

      new AwsCustomResource(this, "ClusterLogsEnabler", {
        policy: AwsCustomResourcePolicy.fromSdkCalls({
          resources: [`${this.cluster.clusterArn}/update-config`],
        }),
        onUpdate: {
          physicalResourceId: { id: `${this.cluster.clusterArn}/LogsEnabler` },
          service: "EKS",
          action: "updateClusterConfig",
          region: this.region,
          parameters: {
            name: this.cluster.clusterName,
            logging: {
              clusterLogging: [
                {
                  enabled: true,
                  types: enabled_types,
                },
                {
                  enabled: false,
                  types: disabled_types,
                }
              ],
            },
          },
        },
        onDelete: {
          physicalResourceId: { id: `${this.cluster.clusterArn}/LogsEnabler` },
          service: "EKS",
          action: "updateClusterConfig",
          region: this.region,
          parameters: {
            name: this.cluster.clusterName,
            logging: {
              clusterLogging: [
                {
                  enabled: false,
                  types: SUPPORTED_CONTROL_PLANE_LOGGING,
                },
              ],
            },
          },
        },
      });
    }
  }
```

It was tested in the following scenarios:

1. nothing is provided as `controlPlaneLogging` -> nothing touched, all the Loggings remain disabled
2. Enable all the loggings
```ts
      controlPlaneLogging: {
        enabled: true,
      },
```
3. Let user choose logging types to enable -> only user specified loggins enabled
```ts
      controlPlaneLogging: {
        enabled: true,
        types: ['api', 'audit'],
      },
```
4. Enable all or part of the loggings, then disable/completely remove the `controlPlaneLogging` reference -> disables all the loggings and put backs the default `Disabled` state
```ts
      controlPlaneLogging: {
        enabled: true,
      },
```

followed by completely removing the `controlPlaneLogging` reference or:
```ts
      controlPlaneLogging: {
        enabled: false,
      },
```

Looking forward for your response, cheers!
